### ISSUE TITLE
clang-tidy 13 on Windows fails with MSVC -external:I flag

### ISSUE THREAD

**Issue by ghost** [llvm/llvm-project#52687]
This issue was originally reported by Nin Loot here: https://gitlab.kitware.com/cmake/cmake/-/issues/22979. Quoting from his bug report:

We recently tried to upgrade from cmake version 3.20 to 3.22 and our builds fail when clang-tidy is being used. The version of clang-tidy is version 13.
The issue is that cmake 3.22 now passes the `-external:I` flag to the MS Compiler and clang-tidy doesn't know that flag so all external headers are not found. I don't have a workaround for this issue yet.
In additional it appears that clang-tidy isn't accepting the `/std:c++17` flag that is passed to the MS Compiler so the wrong version of C++ is used (this is probably similar to issue https://gitlab.kitware.com/cmake/cmake/-/issues/22767 but our invocation of the MSVC compiler command line has the `/std:c++17` so it's not exactly the same issue). This issue was easy to fix by adding `--extra-arg=/std:c++17` to the `CMAKE_CXX_CLANG_TIDY` variable. This is the same solution we used to fix the issue that the flag `/EHsc` wasn't passed to clang-tidy (https://gitlab.kitware.com/cmake/cmake/-/issues/20512).

**Comment by Chilledheart** [comment]
Tried with chromium's main branch of llvm  (tagged `llvmorg-14-init-12719-gc4b45eeb-2` rolled in https://chromium.googlesource.com/chromium/src/+/76b2f19d874cddd74599380979203bcde2b68632) the issue still there.

**Comment by Chilledheart** [comment]
Took a bit at `stripPositionalArgs` func in `lib/Tooling/CompilationDatabase.cpp`.  It seems wrong that it is a correct way to insert `--extra-arg-before=--driver-mode=cl` into the flags because in this function all cl-style flags will be converted into gcc-style flags due to the face all cl-style flags are aliases to gcc-style flags used in clang internal (look at '-external:I" field of  https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Driver/Options.td#L6076 for example).

In algorithm, all compliation flags are translated from `CommonOptionsParser::create` in ClangTidyMain.cpp, `FixedCompilationDatabase::loadFromCommandLine` in CommonOptionsParser.cpp, and finally `stripPositionalArgs` in CompilationDatabase.cpp. Then the flags contained in extra-arg and extra-arg-before are added into the translated compilations flags. There is no way to accept cl-style flags in these two options because all translated flags are already gcc-style and impossible to re-convert them back to cl-style progmatically.

FYI I Patched `stripPositionalArgs` a bit to make it clear:
```
diff --git a/clang/lib/Tooling/CompilationDatabase.cpp b/clang/lib/Tooling/CompilationDatabase.cpp
index 1e19e68633d2..a7fd3423fd5f 100644
--- a/clang/lib/Tooling/CompilationDatabase.cpp
+++ b/clang/lib/Tooling/CompilationDatabase.cpp
@@ -283,8 +283,18 @@ static bool stripPositionalArgs(std::vector<const char *> Args,
   if (!Compilation)
     return false;
 
+  llvm::errs() << "Original Args\n";
+  for (const auto&Arg : Compilation->getInputArgs())
+    Arg->print(llvm::errs());
+
+  llvm::errs() << "Translated Args\n";
+  for (const auto&Arg : Compilation->getArgs())
+    Arg->print(llvm::errs());
+
   const driver::JobList &Jobs = Compilation->getJobs();
 
+  Jobs.Print(llvm::errs(), "\n", true);
+
   CompileJobAnalyzer CompileAnalyzer;
 
   for (const auto &Cmd : Jobs) {
```

For the output of `clang-check -analyze "void a(int *x) { if(x){} *x = 47; }" --extra-arg-before=--driver-mode=cl -- /EHsc /W4` are cl-style flags ignored 

```
[3/3] Linking CXX executable bin/clang-check
Original Args
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:0 Values: ['/EHsc']>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:1 Values: ['/W4']>
< Opt:<FlagClass Prefixes:["-"] Name:"c" Group:<GroupClass Name:"<action group>">
>
 Index:2 Values: []>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:3 Values: ['placeholder.cpp']>
Translated Args
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:0 Values: ['/EHsc']>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:1 Values: ['/W4']>
< Opt:<FlagClass Prefixes:["-"] Name:"c" Group:<GroupClass Name:"<action group>">
>
 Index:2 Values: []>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:3 Values: ['placeholder.cpp']>
< Opt:<JoinedClass Prefixes:["-"] Name:"mlinker-version=">
 Index:4 Values: ['711']>
 "" "-cc1" "-triple" "arm64-apple-macosx12.0.0" "-Wundef-prefix=TARGET_OS_" "-Werror=undef-prefix" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-emit-obj" "-mrelax-all" "--mrelax-relocations" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "placeholder.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=non-leaf" "-ffp-contract=on" "-fno-rounding-math" "-funwind-tables=2" "-fcompatibility-qualified-id-block-type-checking" "-fvisibility-inlines-hidden-static-local-var" "-target-cpu" "apple-m1" "-target-feature" "+v8.5a" "-target-feature" "+fp-armv8" "-target-feature" "+neon" "-target-feature" "+crc" "-target-feature" "+crypto" "-target-feature" "+dotprod" "-target-feature" "+fp16fml" "-target-feature" "+ras" "-target-feature" "+lse" "-target-feature" "+rdm" "-target-feature" "+rcpc" "-target-feature" "+zcm" "-target-feature" "+zcz" "-target-feature" "+fullfp16" "-target-feature" "+sha2" "-target-feature" "+aes" "-target-abi" "darwinpcs" "-fallow-half-arguments-and-returns" "-debugger-tuning=lldb" "-target-linker-version" "711" "-fcoverage-compilation-dir=/Users/ko/llvm-project/llvm/build" "-resource-dir" "lib/clang/14.0.0" "-stdlib=libc++" "-internal-isystem" "/usr/local/include" "-internal-isystem" "lib/clang/14.0.0/include" "-internal-externc-isystem" "/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir=/Users/ko/llvm-project/llvm/build" "-ferror-limit" "19" "-stack-protector" "1" "-fblocks" "-fencode-extended-block-signature" "-fregister-global-dtors-with-atexit" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "placeholder.o" "-x" "c++" "placeholder.cpp"
warning: '/Users/ko/llvm-project/llvm/build/void a(int *x) { if(x){} *x = 47; }' treated as the '/U' option [-Wslash-u-filename]
note: use '--' to treat subsequent arguments as filenames
error: no input files
error: unable to handle compilation, expected exactly one compiler job in ''
Error while processing /Users/ko/llvm-project/llvm/build/void a(int *x) { if(x){} *x = 47; }.
```

While for the output of `clang-check -analyze "void a(int *x) { if(x){} *x = 47; }" -- --driver-mode=cl /EHsc /W4 -external:I.` are cl-style flags not 

```
Original Args
< Opt:<JoinedClass Prefixes:["--"] Name:"driver-mode=" Group:<GroupClass Name:"<clang driver internal options>" Group:<GroupClass Name:"<clang internal options>">
>
>
 Index:0 Values: ['cl']>
< Opt:<JoinedClass Prefixes:["/", "-"] Name:"EH" Group:<GroupClass Name:"<clang-cl options>">
>
 Index:1 Values: ['sc']>
< Opt:<FlagClass Prefixes:["-"] Name:"WCL4" Group:<GroupClass Name:"<W group>" Group:<GroupClass Name:"<W/R group>" Group:<GroupClass Name:"<CompileOnly group>">
>
>
>
 Index:2 Values: []>
< Opt:<JoinedOrSeparateClass Prefixes:["-"] Name:"isystem" Group:<GroupClass Name:"<clang i group>" Group:<GroupClass Name:"<i group>" Group:<GroupClass Name:"<I/i group>" Group:<GroupClass Name:"<Preprocessor group>" Group:<GroupClass Name:"<CompileOnly group>">
>
>
>
>
>
 Index:3 Values: ['.']>
< Opt:<FlagClass Prefixes:["-"] Name:"c" Group:<GroupClass Name:"<action group>">
>
 Index:4 Values: []>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:5 Values: ['placeholder.cpp']>
Translated Args
< Opt:<JoinedClass Prefixes:["--"] Name:"driver-mode=" Group:<GroupClass Name:"<clang driver internal options>" Group:<GroupClass Name:"<clang internal options>">
>
>
 Index:0 Values: ['cl']>
< Opt:<JoinedClass Prefixes:["/", "-"] Name:"EH" Group:<GroupClass Name:"<clang-cl options>">
>
 Index:1 Values: ['sc']>
< Opt:<FlagClass Prefixes:["-"] Name:"WCL4" Group:<GroupClass Name:"<W group>" Group:<GroupClass Name:"<W/R group>" Group:<GroupClass Name:"<CompileOnly group>">
>
>
>
 Index:2 Values: []>
< Opt:<JoinedOrSeparateClass Prefixes:["-"] Name:"isystem" Group:<GroupClass Name:"<clang i group>" Group:<GroupClass Name:"<i group>" Group:<GroupClass Name:"<I/i group>" Group:<GroupClass Name:"<Preprocessor group>" Group:<GroupClass Name:"<CompileOnly group>">
>
>
>
>
>
 Index:3 Values: ['.']>
< Opt:<FlagClass Prefixes:["-"] Name:"c" Group:<GroupClass Name:"<action group>">
>
 Index:4 Values: []>
< Opt:<InputClass Prefixes:[] Name:"<input>">
 Index:5 Values: ['placeholder.cpp']>
< Opt:<JoinedClass Prefixes:["-"] Name:"mlinker-version=">
 Index:9 Values: ['711']>
 "" "-cc1" "-triple" "arm64-pc-windows-msvc19.14.0" "-emit-obj" "-mrelax-all" "-mincremental-linker-compatible" "--mrelax-relocations" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "placeholder.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=none" "-relaxed-aliasing" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "generic" "-target-feature" "+neon" "-target-feature" "+v8a" "-target-abi" "aapcs" "-fallow-half-arguments-and-returns" "-D_MT" "-flto-visibility-public-std" "--dependent-lib=libcmt" "--dependent-lib=oldnames" "-stack-protector" "2" "-fcxx-exceptions" "-fexceptions" "-fexternc-nounwind" "-fdiagnostics-format" "msvc" "-target-linker-version" "711" "-fcoverage-compilation-dir=/Users/ko/llvm-project/llvm/build" "-resource-dir" "lib/clang/14.0.0" "-isystem" "." "-internal-isystem" "lib/clang/14.0.0/include" "-WCL4" "-fdeprecated-macro" "-fdebug-compilation-dir=/Users/ko/llvm-project/llvm/build" "-ferror-limit" "19" "-fno-use-cxa-atexit" "-fms-extensions" "-fms-compatibility" "-fms-compatibility-version=19.14" "-std=c++14" "-fdelayed-template-parsing" "-faddrsig" "-o" "placeholder.obj" "-x" "c++" "placeholder.cpp"
warning: '/Users/ko/llvm-project/llvm/build/void a(int *x) { if(x){} *x = 47; }' treated as the '/U' option [-Wslash-u-filename]
note: use '--' to treat subsequent arguments as filenames
error: no input files
error: unable to handle compilation, expected exactly one compiler job in ''
Error while processing /Users/ko/llvm-project/llvm/build/void a(int *x) { if(x){} *x = 47; }.
```

I tested `/EHsc` and `/W4`. `-external:I` should operate silimilarly as well.

TLDR, it is probably a cmake bug but not clang-tidy bug.

**Comment by hsattler** [comment]
Repeating the conclusion from https://gitlab.kitware.com/cmake/cmake/-/issues/22979

The issue is `-external:I my/path` (with a space, which does not work) and `-external:Imy/path` (without space, this one works).

**Comment by hihig2001** [comment]
it seems it happens with msvc-cmake project with clang-tidy too.
clang-tidy using msvc-cmake also doesn't understand externals.

if i change `windows-clang.cmake`, `-imsvc ` to `-imsvc`

`set(CMAKE_INCLUDE_SYSTEM_FLAG_${lang} "-imsvc ")`
to
`set(CMAKE_INCLUDE_SYSTEM_FLAG_${lang} "-imsvc")`

it works fine.

but, i'd rather not adjust cmake file directly myself.
i put below code in my project.

```
if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
    if (${CMAKE_INCLUDE_SYSTEM_FLAG_CXX} STREQUAL "-imsvc ") #not fixed yet, current cmake 3.23
        set(CMAKE_INCLUDE_SYSTEM_FLAG_CXX "-imsvc")
    elseif(${CMAKE_INCLUDE_SYSTEM_FLAG_CXX} STREQUAL "-external:I ") #fixed in cmake 3.23
        set(CMAKE_INCLUDE_SYSTEM_FLAG_CXX "-external:I")
    endif()
endif()
```

**Comment by PiotrZSL** [comment]
Thing is that clang-tidy does not support --driver-mode in extra args in stripPositionalArgs, and due to that some arguments are striped.
